src/directory/file_slice.rs

```diff
@@ -57,15 +57,18 @@ pub struct FileSlice {
 
 impl FileSlice {
     /// Wraps a FileHandle.
-    pub fn new<D>(data: D) -> Self
-    where
-        D: FileHandle,
-    {
-        let len = data.len();
+    pub fn new(file_handle: Box<dyn FileHandle>) -> Self {
+        let num_bytes = file_handle.len();
+        FileSlice::new_with_num_bytes(file_handle, num_bytes)
+    }
+
+    /// Wraps a FileHandle.
+
+    pub fn new_with_num_bytes(file_handle: Box<dyn FileHandle>, num_bytes: usize) -> Self {
         FileSlice {
-            data: Arc::new(Box::new(data)),
+            data: Arc::new(file_handle),
```

Comment:
I don't know if that buys us anything but we could define `FileSlice` as:
```rust
pub struct FileSlice {
    data: Arc<dyn FileHandle>,
...
```

and write:
```rust
data: Arc::from(file_handle),
```
here

Reply:
excellent point!